packages/gatsby-transformer-documentationjs/src/gatsby-node.js

```diff
@@ -80,76 +60,166 @@ exports.onCreateNode = async ({
-
-      // Defaults
-      picked.params = [{ name: ``, type: { type: ``, name: `` } }]
-      picked.returns = [{ type: { type: ``, name: `` } }]
-      picked.examples = [{ raw: ``, highlighted: `` }]
-
-      // Prepare various sub-pieces.
-      if (docsJson.description) {
-        picked.description___NODE = createDescriptionNode(
-          node,
-          commentId(node.id, i),
-          stringifyMarkdownAST(docsJson.description),
-          `comment.description`,
-          actions,
-          createNodeId
-        )
+    const handledDocs = new WeakMap()
+    const typeDefs = new Map()
+
+    const getNodeIDForType = typeName => {
+      if (typeDefs.has(typeName)) {
+        return typeDefs.get(typeName)
       }
 
-      const transformParam = param => {
-        if (param.description) {
-          param.description___NODE = createDescriptionNode(
-            node,
-            commentId(node.id, i),
-            stringifyMarkdownAST(param.description),
-            param.name,
-            actions,
-            createNodeId
-          )
-          delete param.description
-        }
-        delete param.lineNumber
-
-        // When documenting destructured parameters, the name
-        // is parent.child where we just want the child.
-        if (param.name.split(`.`).length > 1) {
-          param.name = param.name
-            .split(`.`)
-            .slice(-1)
-            .join(`.`)
-        }
+      const index = documentationJson.findIndex(
+        docsJson =>
+          docsJson.name === typeName &&
+          [`typedef`, `constant`].includes(docsJson.kind)
+      )
 
-        if (param.properties) {
-          param.properties = param.properties.map(transformParam)
-        }
+      if (index !== -1) {
+        return prepareNodeForDocs(documentationJson[index], {
+          commentNumber: index,
+        }).node.id
+      }
+
+      return null
+    }
 
-        return param
+    const tryToAddTypeDef = type => {
+      if (type.applications) {
+        type.applications.forEach(tryToAddTypeDef)
       }
 
-      if (docsJson.params) {
-        picked.params = docsJson.params.map(transformParam)
+      if (type.expression) {
+        tryToAddTypeDef(type.expression)
       }
 
-      if (docsJson.returns) {
-        picked.returns = docsJson.returns.map(ret => {
-          if (ret.description) {
-            ret.description___NODE = createDescriptionNode(
-              node,
-              commentId(node.id, i),
-              stringifyMarkdownAST(ret.description),
-              ret.title,
-              actions,
-              createNodeId
-            )
-            delete ret.description
-          }
+      if (type.elements) {
+        type.elements.forEach(tryToAddTypeDef)
+      }
 
-          return ret
-        })
+      if (type.type === `NameExpression` && type.name) {
+        type.typeDef___NODE = getNodeIDForType(type.name)
+      }
+    }
+
+    const prepareNodeForDocs = (
```

Comment:
maybe add a few comments to this function to tell you what you're doing as from reading the code I can't but I have no knowledge of parsing documentation 😄

Reply:
Added in https://github.com/gatsbyjs/gatsby/pull/11597/commits/2c055f85ca11ede26095a6510605edab956df20b (there is separate PR for documentation plugin). Commits here are only so netlify preview use those changes